contracts/abstracted_account.algo.ts

```diff
@@ -35,35 +42,169 @@ export class AbstractedAccount extends Contract {
+
+    assert(this.plugins(key).exists, 'plugin does not exist');
+    assert(this.plugins(key).value.lastValidRound >= globals.round, 'plugin is expired');
+    assert(
+      globals.round - this.plugins(key).value.lastCalled >= this.plugins(key).value.cooldown,
+      'plugin is on cooldown'
+    );
+  }
+
+  private pluginCallAllowed(app: AppID, caller: Address): boolean {
+    const key: PluginsKey = { application: app, allowedCaller: caller };
+
+    return (
+      this.plugins(key).exists &&
+      this.plugins(key).value.lastValidRound >= globals.round &&
+      globals.round - this.plugins(key).value.lastCalled >= this.plugins(key).value.cooldown
+    );
+  }
+
+  private ensuresRekeyBack(txn: Txn): boolean {
+    if (txn.sender === this.app.address && txn.rekeyTo === this.app.address) {
+      return true;
+    }
+
+    return (
+      txn.typeEnum === TransactionType.ApplicationCall &&
+      txn.applicationID === this.app &&
+      txn.numAppArgs === 1 &&
+      txn.onCompletion === 0 &&
+      txn.applicationArgs[0] === method('arc58_verifyAuthAddr()void')
+    )
+  }
+
+  private assertRekeysBack(): void {
+    let rekeysBack = false;
+    for (let i = (this.txn.groupIndex + 1); i < this.txnGroup.length; i += 1) {
+      const txn = this.txnGroup[i];
+
+      if (this.ensuresRekeyBack(txn)) {
+        rekeysBack = true;
+      }
+    }
+
+    assert(rekeysBack, 'rekey back not found');
+  }
+
   /**
-   * Ensure that by the end of the group the abstracted account has control of itself
+   * Guarantee that our txn group is valid in a single loop over all txns in the group
+   * 
+   * @param app the plugin app id being validated
+   * @param checkGlobal whether to check the global caller for method restrictions
+   * @param checkLocal whether to check the local caller for method restrictions
    */
-  private verifyRekeyToAbstractedAccount(): void {
-    let rekeyedBack = false;
+  private assertValidGroup(app: AppID, methodOffsets: uint64[], checkGlobal: boolean, checkLocal: boolean): CallerUsed {
+    const gkey: PluginsKey = { application: app, allowedCaller: Address.zeroAddress };
+    const key: PluginsKey = { application: app, allowedCaller: this.txn.sender };
+
+    let rekeysBack = false;
+
+    const globalRestrictions = checkGlobal && this.plugins(gkey).size > 29;
+    const localRestrictions = checkLocal && this.plugins(key).size > 29;
+    let methodIndex = 0;
+    let callerUsed: CallerUsed = {
+      global: checkGlobal && !globalRestrictions,
+      local: checkLocal && !localRestrictions,
+    };
 
-    for (let i = this.txn.groupIndex + 1; i < this.txnGroup.length; i += 1) {
+    for (let i = (this.txn.groupIndex + 1); i < this.txnGroup.length; i += 1) {
       const txn = this.txnGroup[i];
 
-      // The transaction is an explicit rekey back
-      if (txn.sender === this.app.address && txn.rekeyTo === this.app.address) {
-        rekeyedBack = true;
-        break;
+      if (this.ensuresRekeyBack(txn)) {
```

Comment:
Just noting that this will end up using more ops because the `callsub... proto... retsub...` for each call, but I imagine this will be re-implemented in PuyaTS by the time the ARC is finalized which will support in-lining

Reply:
noted. I think we'll be fine, i agree this likely will end up being puya-ts before finalization and it'll be inlined